### PR TITLE
Added description of using samples from OptimizationResult.

### LINKED ISSUE #82: Add a tutorial to describe how to use samples of the optimization result
### What is the expected enhancement?

The optimization result includes not only the best solution but also solution samples.
It would be nice to write a tutorial to introduce how to use samples, e.g., compute statistics such as mean, percentile, standard deviation and visualize a histogram #81.

### PR DESCRIPTION
**Pull request by MartinBeseda**
This commit contributes to the solution of issue #82.

### Summary
_Analysis_ section of docs/tutorials/10_warm_start_qaoa.ipynb was rewritten to contain more in-depth information about how to use samples extracted from `OptimizationResult` object and perform basic statistics over them. This is according to the request in issue #82.

Closes #82

### CODE REVIEW

**Comment by t-imamichi** [comment]
I think it's better to separate a tutorial of solution samples with Warm Start QAOA because solution samples are not limited for Warm Start QAOA. Could you make a separate tutorial with a simple QAOA?

**Comment by t-imamichi** [comment]
Could you revert or copy of the original 10_warm_start_qaoa.ipynb? You don't have to modify it because you write about samples in section 3.

File: docs/tutorials/03_minimum_eigen_optimizer.ipynb [hunk 1]
```diff
@@ -293,6 +297,270 @@
...
+   }
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 48,
+   "outputs": [],
+   "source": [
+    "def print_samples(samples: List[SolutionSample],\n",
+    "                  max_len: int = None):\n",
+    "    status_str = {OptimizationResultStatus.SUCCESS: 'SUCCESS',\n",
```
Comment (t-imamichi) [inline docs/tutorials/03_minimum_eigen_optimizer.ipynb:322]: you don't need to create a dict. `Enum.name` is enough. `sample.name`

File: docs/tutorials/03_minimum_eigen_optimizer.ipynb [hunk 2]
```diff
@@ -293,6 +297,270 @@
...
+    "                  OptimizationResultStatus.FAILURE: 'FAILURE',\n",
+    "                  OptimizationResultStatus.INFEASIBLE: 'INFEASIBLE'}\n",
+    "    print('\\n'.join(['[{}]: fval: {:.4f}, probability: {:.2f} [{}]'.format(''.join(str(int(_)) for _ in e.x),\n",
+    "                                                                           e.fval,\n",
+    "                                                                           e.probability,\n",
+    "                                                                           status_str[e.status])\n",
+    "            for e in sorted(samples, key=lambda x: x.probability, reverse=True)][:max_len]))"
```
Comment (t-imamichi) [inline docs/tutorials/03_minimum_eigen_optimizer.ipynb:329]: Please do not sort the samples by yourself here. qiskit-opt sorts the samples by the order  of the objective function value.

Comment (t-imamichi) [reply docs/tutorials/03_minimum_eigen_optimizer.ipynb:329]: I don't think you need to define a function. The following snippet is enough.
```
for sample in samples:
    print(sample)
```

File: docs/tutorials/03_minimum_eigen_optimizer.ipynb [hunk 3]
```diff
@@ -293,6 +297,270 @@
...
+    "collapsed": false
+   }
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 50,
+   "outputs": [],
+   "source": [
+    "def get_filtered_samples(samples: List[SolutionSample],\n",
+    "                         percentile: float = 0,\n",
```
Comment (t-imamichi) [inline docs/tutorials/03_minimum_eigen_optimizer.ipynb:383]: `percentile` should be `threshold`. Here it does not compute percentile. It is just used as a threshold.

File: docs/tutorials/03_minimum_eigen_optimizer.ipynb [hunk 4]
```diff
@@ -293,6 +297,270 @@
...
+   "metadata": {
+    "collapsed": false
+   }
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 52,
+   "outputs": [],
+   "source": [
+    "class PropertyChoice(Enum):\n",
```
Comment (t-imamichi) [inline docs/tutorials/03_minimum_eigen_optimizer.ipynb:447]: You don't need to introduce another Enum for filtering.
You can just generate two list
```
fvals = [sample.fval for sample in result.samples]
probs = [sample.probability for sample in result.samples]
```

**Comment by t-imamichi** [comment]
The examples look good. But they are a bit complicated. You don't have to make a function most cases. Could you simplify the examples?

**Comment by MartinBeseda** [comment]
Dear @t-imamichi ,

thank you very much for your input! I was thinking, that functions will be useful for new Qiskit users, as they can simply copy-paste them, but I understand your objections concerning the simplicity.

I'll make the necessary edits asap.

Best,
Martin

**Comment by t-imamichi** [comment]
Can you update the tutorial? If you cannot, we will do it on behalf of you.

**Comment by MartinBeseda** [comment]
Dear @t-imamichi ,
I'll update the tutorial today or tomorrow at last.

Martin

**Comment by MartinBeseda** [comment]
Dear @t-imamichi ,
I've performed the necessary changes, codes are simplified, functions except `get_filtered_samples()` removed and Tutorial 10 should be the very same version as in the original Qiskit repo.

**Comment by t-imamichi** [comment]
Thanks! It looks quite good. I have two comments.

1. Could you mention the variable order of `x` as follows?
```python
print('variable order:', [var.name for var in qaoa_result.variables])
for s in qaoa_result.samples:
    print(s)
```
output:
```
variable order: ['x', 'y', 'z']
SolutionSample(x=array([0., 1., 0.]), fval=-2.0, probability=0.12499999999999994, status=<OptimizationResultStatus.SUCCESS: 0>)
SolutionSample(x=array([0., 0., 0.]), fval=0.0, probability=0.12499999999999994, status=<OptimizationResultStatus.SUCCESS: 0>)
SolutionSample(x=array([1., 1., 0.]), fval=0.0, probability=0.12499999999999994, status=<OptimizationResultStatus.SUCCESS: 0>)
SolutionSample(x=array([1., 0., 0.]), fval=1.0, probability=0.12499999999999994, status=<OptimizationResultStatus.SUCCESS: 0>)
SolutionSample(x=array([0., 0., 1.]), fval=3.0, probability=0.12499999999999994, status=<OptimizationResultStatus.SUCCESS: 0>)
SolutionSample(x=array([1., 0., 1.]), fval=3.0, probability=0.12499999999999994, status=<OptimizationResultStatus.SUCCESS: 0>)
SolutionSample(x=array([0., 1., 1.]), fval=3.0, probability=0.12499999999999994, status=<OptimizationResultStatus.SUCCESS: 0>)
SolutionSample(x=array([1., 1., 1.]), fval=4.0, probability=0.12499999999999994, status=<OptimizationResultStatus.SUCCESS: 0>)
```

2. Could you use the following snippet to make sure the variable order?
```python
samples_for_plot = {' '.join(f'{qaoa_result.variables[i].name}={int(v)}' for i, v in enumerate(s.x)): s.probability for s in filtered_samples}
samples_for_plot
```
output:
```
{'x=0 y=1 z=0': 0.12499999999999994,
 'x=0 y=0 z=0': 0.12499999999999994,
 'x=1 y=1 z=0': 0.12499999999999994,
 'x=1 y=0 z=0': 0.12499999999999994,
 'x=0 y=0 z=1': 0.12499999999999994,
 'x=1 y=0 z=1': 0.12499999999999994,
 'x=0 y=1 z=1': 0.12499999999999994,
 'x=1 y=1 z=1': 0.12499999999999994}
```

**Comment by MartinBeseda** [comment]
Dear @t-imamichi , 
thank you, I've pushed the changes. 

Best,
Martin

File: docs/tutorials/03_minimum_eigen_optimizer.ipynb [hunk 5]
```diff
@@ -293,6 +296,247 @@
     "print(qaoa_result)"
    ]
   },
+  {
+   "cell_type": "markdown",
+   "source": [
+    "### Analysis of Samples\n",
+    "Objects of type `SolutionSample` (here denoted as *samples*) are a useful information source provided\n",
+    "by `OptimizationResult` objects. They contain information about input values `x`, corresponding (optimized) function\n",
```
Comment (t-imamichi) [inline docs/tutorials/03_minimum_eigen_optimizer.ipynb:304]: corresponding (optimized) function value -> objective function value

File: docs/tutorials/03_minimum_eigen_optimizer.ipynb [hunk 6]
```diff
@@ -293,6 +296,247 @@
...
+    "pycharm": {
+     "name": "#%%\n"
+    }
+   }
+  },
+  {
+   "cell_type": "markdown",
+   "source": [
+    "If we want to obtain a better perspective of the results, statistics is very helpful, both with respect to\n",
+    "optimized function values and their respective probabilities. Thus, mean and standard deviation are the very basics for\n",
```
Comment (t-imamichi) [inline docs/tutorials/03_minimum_eigen_optimizer.ipynb:413]: optimized function values -> the objective function values

File: docs/tutorials/03_minimum_eigen_optimizer.ipynb [hunk 7]
```diff
@@ -293,6 +296,247 @@
     "print(qaoa_result)"
    ]
   },
+  {
+   "cell_type": "markdown",
+   "source": [
+    "### Analysis of Samples\n",
+    "Objects of type `SolutionSample` (here denoted as *samples*) are a useful information source provided\n",
```
Comment (t-imamichi) [inline docs/tutorials/03_minimum_eigen_optimizer.ipynb:303]: I prefer writing "`OptimizationResult` provides a useful information source `SolutionSample` (here denoted as *samples*)."

File: docs/tutorials/03_minimum_eigen_optimizer.ipynb [hunk 8]
```diff
@@ -293,6 +296,247 @@
...
+    "collapsed": false,
+    "pycharm": {
+     "name": "#%%\n"
+    }
+   }
+  },
+  {
+   "cell_type": "markdown",
+   "source": [
+    "We may also want to filter samples according to their status or to a certain percentile, i.e. with probability higher\n",
```
Comment (t-imamichi) [inline docs/tutorials/03_minimum_eigen_optimizer.ipynb:347]: according to their status or to a certain percentile, i.e., ... -> according to their status or probabilities. (omit i.e. part)

**Comment by t-imamichi** [comment]
Thank you for the revision. I have just a few minor comments. It's almost ready to merge. Could you "restart and run all" to reset cell numbers when you push the final version?

**Comment by MartinBeseda** [comment]
Dear @t-imamichi , thank you for the suggestions, I've completed your request. 

Best,
Martin

File: docs/tutorials/03_minimum_eigen_optimizer.ipynb [hunk 9]
```diff
@@ -350,14 +591,67 @@
...
+   "cell_type": "code",
+   "execution_count": 20,
+   "metadata": {
+    "pycharm": {
+     "name": "#%%\n"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "samples_for_plot = {''.join(str(int(_)) for _ in s.x): s.probability for s in filtered_samples}"
```
Comment (t-imamichi) [inline docs/tutorials/03_minimum_eigen_optimizer.ipynb:617]: Could you update the keys of the dictionary with variable names as Cell [15]?

**Comment by t-imamichi** [comment]
Thank you for the revision. I'm afraid that I missed a minor point of the plot for RQAOA. This will be the final change...

**Comment by MartinBeseda** [comment]
Dear @t-imamichi , 
thank you again - I've changed the code accordingly.

Martin

**Review by t-imamichi — APPROVED** [review]
LGTM. Thanks!

**Comment by MartinBeseda** [comment]
Dear @t-imamichi , thank you for your collaboration!

Martin

**Pull request merged**